// @sentry/wizard/dist/lib/__tests__/Setup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
jest.mock('../Helper/Logging'); // We mock logging to not pollute the output
var Constants_1 = require("../Constants");
var Setup_1 = require("../Setup");
describe('Wizard', function () {
    describe('React Native', function () {
        test('run', function () {
            expect(Setup_1.run({
                quiet: true,
                integration: Constants_1.Integration.reactNative,
                platform: [Constants_1.Platform.ios, Constants_1.Platform.android],
                skipConnect: true,
            })).toBeTruthy();
        });
    });
    describe('Cordova', function () {
        test('run', function () {
            expect(Setup_1.run({
                quiet: true,
                integration: Constants_1.Integration.cordova,
                platform: [Constants_1.Platform.ios, Constants_1.Platform.android],
                skipConnect: true,
            })).toBeTruthy();
        });
    });
    describe('Electron', function () {
        test('run', function () {
            expect(Setup_1.run({
                quiet: true,
                integration: Constants_1.Integration.electron,
                skipConnect: true,
            })).toBeTruthy();
        });
    });
    describe('JavaScript', function () {
        test('run', function () {
            expect(Setup_1.run({
                quiet: true,
                integration: Constants_1.Integration.javascript,
                skipConnect: true,
            })).toBeTruthy();
        });
    });
    describe('Node', function () {
        test('run', function () {
            expect(Setup_1.run({
                quiet: true,
                integration: Constants_1.Integration.node,
                skipConnect: true,
            })).toBeTruthy();
        });
    });
});
